{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.4"},"toc":{"base_numbering":1,"nav_menu":{},"number_sections":true,"sideBar":true,"skip_h1_title":false,"title_cell":"Table of Contents","title_sidebar":"Contents","toc_cell":false,"toc_position":{"height":"calc(100% - 180px)","left":"10px","top":"150px","width":"165px"},"toc_section_display":true,"toc_window_display":true},"varInspector":{"cols":{"lenName":16,"lenType":16,"lenVar":40},"kernels_config":{"python":{"delete_cmd_postfix":"","delete_cmd_prefix":"del ","library":"var_list.py","varRefreshCmd":"print(var_dic_list())"},"r":{"delete_cmd_postfix":") ","delete_cmd_prefix":"rm(","library":"var_list.r","varRefreshCmd":"cat(var_dic_list()) "}},"oldHeight":163.489666,"position":{"height":"184.823px","left":"718px","right":"20px","top":"75px","width":"405.99px"},"types_to_exclude":["module","function","builtin_function_or_method","instance","_Feature"],"varInspector_section_display":"block","window_display":false},"colab":{"name":"Intro_to_Python_post_class_version.ipynb","provenance":[{"file_id":"1Szly4gUhuaChB1vX8AAud4YLA-EM3iOk","timestamp":1650476095704},{"file_id":"18raJW2yNykpfKVxEAlNuvkzMFtB2VBD1","timestamp":1596664819628},{"file_id":"1s7V_vIzy3WckbVbenOAfx65LKscKmeRN","timestamp":1591737599774}],"collapsed_sections":["CJKaTVe7zlRV","eXAv5w3UPMoJ","z5pYjZcDPMo7","JDJ5twWF7cRO","ytQEWGVd7kEw"]}},"cells":[{"cell_type":"markdown","metadata":{"id":"CJKaTVe7zlRV"},"source":["# <center>**Introduction to Python Workshop**</center>\n","###### <center>[Harvard Chan Bioinformatics Core](https://bioinformatics.sph.harvard.edu/)</center>\n","###### <center>2022-8-10</center>\n"]},{"cell_type":"markdown","metadata":{"id":"1b6cELGTPiUX"},"source":["---\n","\n","# Pre-class preparation (10 min)\n","Welcome to **Introduction to Python** workshop! Before we start, let's prepare a few things and get familiar with the platform (Google Colab) that we will use during this workshop. Colab is a free cloud service platform based on the [Jupyter notebook](https://jupyter.org/) environment. It does not require users to install python locally, and the code is run entirely on the cloud server.\n","\n","Please follow the instructions below to set up for this workshop:\n","1. **Download the [data](https://github.com/hbctraining/Training-modules/blob/master/Python/data_python_workshop.zip?raw=true) needed for this workshop**: Please save the file and unzip it on your computer (e.g. desktop).\n","2. **Create your own copy of this python notebook (the workshop materials)**: Click \"File\" at the menu bar -> Click \"Save a copy in Drive\". This notebook will be automatically named \"Copy of Intro_to_Python_in_class_version\" and copied in your Google Drive. You can rename it if you want.\n","3. **Double check where the material is located**: Click \"File\" again -> click \"Locate in Drive\". A new window should pop up, showing you its location in your Google Drive.\n","4. **Get familiar with the Colab user interface and terminology**: \n","*   code cell\n","*   text cell\n","*   creating and deleting cell\n","*   adding comments\n","*   saving the document\n"]},{"cell_type":"markdown","metadata":{"id":"qHPTgwHWPMn2"},"source":["---\n","\n","# Section 1: Introduction to Python (10 min)\n","\n","## What is Python?\n","Python is a powerful, open-source, general-purpose programming language with a wide variety of applications. Many [websites and apps](https://codeinstitute.net/blog/7-popular-software-programs-written-in-python/) that we are familiar with, for example YouTube, Instagram, Spotify etc, are built on Python. In the field of bioinformatics, Python is also widely used in computational programming, data analysis, and pipeline development. \n","\n","Below we have listed some examples where Python-based tools are used:\n","\n","| **Example use case** | **Tool** |\n","| :---: | :---: |\n","| Pipeline development | [bcbio](https://bcbio-nextgen.readthedocs.io/en/latest/), [snakemake](https://snakemake.readthedocs.io/en/stable/) |\n","| Image analysis | [CellProfiler](https://cellprofiler.org/) |\n","| Molecular visualization | [PyMOL](https://pymol.org/2/) |\n","| Machine learning | [scikit-learn](https://scikit-learn.org/stable/) |\n","\n","> Note: While many bioinformatics tools are written in Python, you may have used them (for example, through the command line) without knowing how they work internally. This is okay, but if you want to modify the script for customized usage, it is better to learn the syntax of the script.\n","\n","Since it was first released in 1991, Python has been constantly developed and updated. The current version of Python is 3.0. Python programming language features include:\n","- Easy-to-understand syntax\n","- Large number of built-in libraries for common tasks\n","- Succinct and readable format, friendly to both programming amateur and professional\n","- Widely used programming language, with comprehensive documentation and plenty of tutorials\n","\n","## What will you learn in this workshop?\n","To take this workshop, you don't need to have prior experience in Python (or any other programming language(s)). We will learn the basic Python syntax, data structure, function, data manipulation, and data visulization. At the end of the workshop, students should be able to understand the structure of a simple Python script, and can perform basic data analysis.\n","\n","## Additional tips\n","- `#` at the beginning of a line denotes that this line is a comment for the code - any line starting with a `#` won't be executed by Python. \n","- Indentation is very important in Python syntax. Not only is it necessary for Python, it is also great for code readability. By making indentations essential, Python forces good code-writing."]},{"cell_type":"markdown","metadata":{"id":"2XlURgD6PMn3"},"source":["---\n","\n","# Section 2: Basic Python syntax (20 min)"]},{"cell_type":"markdown","metadata":{"id":"MoK64s_g6BTo"},"source":["## Variables\n","A \"variable\" is a temporary container that stores certain information. We use the `=` operator to assign some value to a variable. The naming of a variable in Python has to fulfill the following rules:\n","- must start with a letter or underscore character\n","- can contain only alpha-numeric characters or the underscore character (A-z, 0-9, _ )\n","- cannot be a reserved keyword in Python. A complete list can be found [here](https://docs.python.org/3.8/reference/lexical_analysis.html#keywords).\n","- is case-sensitive (e.g. `Year` and `year` are two different variable names)  "]},{"cell_type":"code","metadata":{"id":"eWPzzdRQPMn3"},"source":["# Assign two variables: 2 to x, and 5 to y\n","x = 2\n","y = 5"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"d5gnc0lcPMn6"},"source":["The \"x\" and \"y\" variables are now stored in the current Python computing environment. We can see the variables in the variable inspector, `{x}`, located at the left side panel.\n","\n","We could also use the `print()` function, to print out the value of variables to the console. Alternatively, we could just type out the variable name. \n","> If running mutliple lines of code in a code chunk, it is recommended to use the `print()` function. Otherwise, only the value of the last variable in your code chunk will be printed out to the console."]},{"cell_type":"code","metadata":{"id":"NL01057zPMn7"},"source":["# Print out the value of variables 'x' and 'y'\n","print(x)\n","print(y)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Kd3r30rpZFRW"},"source":["> Note: Within a line of code, Python is lenient about whether to use spaces or not. Using space could help readability, but it is a personal preference."]},{"cell_type":"markdown","metadata":{"id":"swFsV8xjPMn-"},"source":["New variables can be generated by performing some mathematical calculations on existing variables. For example, we can calculate the mean of x and y. \n","\n","Python also supports a wide range of common mathematical calculations using operators and functions, e.g. power (`**`) and square root (`sqrt()`)."]},{"cell_type":"code","metadata":{"id":"Bwz1WawIPMn_"},"source":["# Calculate the mean of x and y, and store the value to a variable called 'mean'\n","mean = (x + y) / 2\n","mean"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"63BaOVc4PMoB"},"source":["## Data types\n","Data comes in different types. For example, the newly created variables `x` and `mean` are numeric. `x` is a whole number, so its data type is `int`, or \"integer\"; `mean`, on the other hand, is a number with decimal places, so its data type is `float`. \n","\n","> These are similar to the \"integer\" and \"numeric\" data types in R, respectively.\n","\n","We can use the `type()` function to check what data type a given variable has."]},{"cell_type":"code","metadata":{"id":"JDKPk7l6PMoB"},"source":["# Check the data type for 'x' and 'mean'\n","print(type(x))\n","print(type(mean))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"N7BJ9miYPMoE"},"source":["Another commonly used data type is `str`. String stores a sequence of characters, and can be created by enclosing characters inside single quotation marks `''` or double quotation marks `\"\"` .\n","\n","> This is similar to the \"character\" data type in R."]},{"cell_type":"code","metadata":{"id":"AkpcdOA7PMoE"},"source":["# Generate a str variable called 'text', with the value 'hello world!'. Check its data type\n","text = 'hello world!'\n","print(text)\n","print(type(text))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"cBClvEuRPMoH"},"source":["The last data type we introduce here is `bool`. The boolean data type can be either `True` or `False`. It is to specify if an expression is true or false. We will cover this data type in the conditional statement section.\n","\n","> This is similar to the \"logical\" data type in R."]},{"cell_type":"code","metadata":{"id":"RXFkI9x9PMoH"},"source":["# Generate a boolean variable called 'test', which judges whether 10 is smaller than 8. Check its data type\n","z = 10 < 8\n","print(z)\n","print(type(z))"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"eXAv5w3UPMoJ"},"source":["## Recap\n","In this section, we introduced some basic terms in Python. We learned **how to assign variables**, and what rules to follow. We also described several important **data types** - `int`, `float`, `str`, `bool`. Hope it has been fun so far!\n","\n","| **Data type** | **Examples** |\n","| :---: | :---: |\n","| int (numeric) | 2 |\n","| float (numeric) | 3.5 |\n","| str | 'hello world!' |\n","| bool | True, False|\n","\n","In the next section, we will focus on one important concept - Python lists. This will be something you use all the time in Python."]},{"cell_type":"markdown","metadata":{"id":"HE7482PePMoK"},"source":["---\n","\n","# Section 3: Python List (40 min)"]},{"cell_type":"markdown","metadata":{"id":"BsBcL_stfya4"},"source":["## Create a List\n","Let's talk about data structures. In Python, data is stored in specific ways within variables. A frequently used \"data structure\" is `list`. A Python list is a collection of data stored within a square bracket `[]`.  \n","\n","A list has the following features:\n","- order of its elements matters\n","- can store mixed data types that we introduced above\n","- can even contain a sublist\n","\n","> Note: There are other Python data structures, including `tuple`, `dictionary`, `sets`. We will not cover those in this workshop, but they can be very useful in some situations. If you are interested in learning more about them, this [website](https://thomas-cokelaer.info/tutorials/python/data_structures.html) has more information."]},{"cell_type":"code","metadata":{"id":"t6gedZQUPMoK"},"source":["# Create an empty list called 'empty'\n","empty = []\n","empty"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"GWhrsGlxPMoM"},"source":["# Create a list called 'species', containing three strings: ecoli, human, corn.\n","species = ['ecoli', 'human', 'corn']\n","species"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"fnBRFSXMPMoP"},"source":["# Create a list called 'glengths', containing three numeric values that correponds to genome length (in Mb): 4.6, 3000, 2500\n","glengths = [4.6, 3000, 2500]\n","glengths"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"UOlOW86_PMoR"},"source":["# Create a list called 'combined', containing all three species and corresponding genome lengths as pairs\n","combined = ['ecoli', 4.6, 'human', 3000, 'corn', 2500]\n","combined"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"jfm-2lHjPMoT"},"source":["# Create a list called 'combined2', with each species and genome length pair as a sublist\n","combined2 = [['ecoli', 4.6],['human', 3000], ['corn', 2500]]\n","combined2"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"2hQbwlbNPMoV"},"source":["## Subsetting a single element from a list\n","Now that we created a list, how do we access the data from it? \n","\n","We can do so by specifying the \"index\" number - the location of the data within the list. **Python index starts from 0** (it is not intuitive, we know! Please just bear with it). \n","\n","The first element of a list is `list[0]`. Alternatively, we can also use `-` to access the data starting from the last element. The last element of a list is `list[-1]`. The image below illustrates the index for each elements. \n","\n","<p align=\"center\">\n","<img src=\"https://github.com/hbctraining/Training-modules/blob/master/Python/img/list1.png?raw=true\" width=\"500\"/>\n","</p>\n","\n"]},{"cell_type":"code","metadata":{"id":"Jcx6ADYdPMoV"},"source":["# Get the 3rd element from list 'combined'\n","combined[2]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"sqrK7FRwPMoX"},"source":["# Get the 3rd element from list 'combined2'. Notice that the result is a sublist!\n","combined2[2]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"i4PoJ2WVPMoZ"},"source":["# Get the 3rd from the last element from the list 'combined'\n","combined[-3]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"_Xdxp8HJPMob"},"source":["## Subsetting multiple elements from a list\n","Now, what if we want to access multiple elements in a list? \n","\n","Here we introduce the slicing `:` operator. The syntax of \"slicing\" is `[start:stop:step]`. *start* refers to the starting index of the slice. *stop* refers to the index of the first element just **after** the finish of our \"slice\". *step* refers to step value of the slice.\n","> Note: You don't have to specify all slicing elements; when it is not specified, Python will use default value - **by default, it will start from the first element, stop at the last element, and use step of 1**.\n","\n","<p align=\"center\">\n","<img src=\"https://github.com/hbctraining/Training-modules/blob/master/Python/img/list2.png?raw=true\" width=\"500\"/>\n","</p>"]},{"cell_type":"code","metadata":{"id":"Z7NvuwFIPMob"},"source":["# Get the first two elements from the list 'combined' - method 1: specify both start and stop position\n","combined[0:2]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"L73ZL1GPPMod"},"source":["# Get the first two elements from the list 'combined' - method 2: specify only stop position\n","combined[:2]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"PikkO1r0PMof"},"source":["# Get the last two elements from the list 'combined' - method 1: use normal index\n","combined[4:]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"wqZPi-ZAPMog"},"source":["# Get the last two elements from the list 'combined' - method 2: use negative index\n","combined[-2:]"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"U9JBRP8KPMoi"},"source":["# Get every other element from the list 'combined'\n","combined[::2]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"z5pYjZcDPMo7"},"source":["## Recap\n","In this section, we introduced **what is a list** and **how to create a list in Python**. We learned **how to access and manipulate one or more elements in a list**. Sometimes there are multiple ways to achieve this goal.\n","\n","In the next section, we are going to introduce tools that make Python programming truly powerful - functions."]},{"cell_type":"markdown","metadata":{"id":"SecRBV8EPMo7"},"source":["---\n","\n","# Section 4: Functions (20 min)"]},{"cell_type":"markdown","metadata":{"id":"tl03J5Kw6BpS"},"source":["## Built-in function\n","A function is a collection of reusable code that performs a particular task. Python has a set of built-in [functions](https://docs.python.org/3/library/functions.html). For example, `max()` returns the maximum value of a list consisting of numeric numbers."]},{"cell_type":"code","metadata":{"id":"cMQ-ZQUzPMo7"},"source":["# Use the max() function to return maximum value of the list 'glengths'\n","max(glengths)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Y0d2MRuBPMo9"},"source":["Let's take another example - `round()` - this function rounds a numeric value to a certain decimal point. By default, the output will be a whole number. "]},{"cell_type":"code","metadata":{"id":"E6QGrRxWPMo-"},"source":["# define a variable with the value of pi, and then output the corresponding whole number using the round() function \n","pi = 3.14159\n","round(pi)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Dey3soYxPMpA"},"source":["What if we want to round the value to specific number of decimal places? In that case, we would have to use additional *arguments* when using the function. \n","\n","To check the available arguments and usage information for a function, one can use the `help()` function. However, we would recommend that you search the web for the function you want to use. For instance, [this webpage](https://www.programiz.com/python-programming/methods/built-in/round) shows some nice examples for the `round()` function. You can easily find similar resources online for most other functions."]},{"cell_type":"code","metadata":{"id":"2wkU77wAPMpA"},"source":["# Use the `help()` function to display the usage of the `round()` function\n","help(round)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"LyhJL_N4PMpC"},"source":["We now know that we can specify number of digits using the `ndigits` argument within `round()`. Let's try that with pi!"]},{"cell_type":"code","metadata":{"id":"PaeNiNP5PMpC"},"source":["# round the value of pi to 2 decimal places\n","round(pi, ndigits=2)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"c0LMXbMHPMpE"},"source":["### Exercise\n","Another useful base function is `sorted()` - it sorts the elements of a given list in a specific order. Use this function to reorder the `glengths` list in **descending** order. Check [here](https://www.programiz.com/python-programming/methods/built-in/sorted) if you are not sure what argument to use."]},{"cell_type":"code","metadata":{"id":"LXjGBA1iPMpE"},"source":["# Sort the glengths list in descending order\n","#### Insert your code below ####\n","sorted(glengths, reverse = True)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"zoQz1NUQPMpG"},"source":["## Object-specific function\n","Python has a lot of functionality beyond the basic built-in functions. Recall the data types and data structures that we learned earlier? They are all called Python **objects**. Depending on the object type, there are functions to perform object-specific tasks. \n","\n","Let's take a concrete example. One function for a Python string is `count`. `count` searches the substring in the given string and returns how many times the substring is present within the object. The syntax is `string.count(substring)`."]},{"cell_type":"code","metadata":{"id":"HsVClRHLPMpG"},"source":["# Count number of T in a DNA sequence 'ACTGAT'\n","DNA = \"TCAGTT\"\n","DNA.count(\"T\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"h7fTU9VqPMpK"},"source":["Pretty handy, right? We have just touched the tip of the iceberg so far. There are many more [functions](https://docs.python.org/3/library/stdtypes.html#string-methods) for strings in Python. Below we list a few more functions that you will likely see or may use.\n","\n","| **Function** | **Description** | **Example** | **Output** |\n","| :---: | :---: | :---: | :---: |\n","| capitalize | Converts the first character to upper case | 'atgc'.capitalize() | 'Atgc' |\n","| count | Returns the number of times a specified value occurs in a string | 'atgc'.count('c') | 1 |\n","| islower | Returns True if all characters in the string are lower case | 'atgc'.islower() | True |\n","| join | Joins the elements of an iterable to the end of the string | ''.join(['a', 't', 'g', 'c']) | 'atgc' |\n","| replace | Returns a string where an old value is replaced with a new value | 'atgc'.replace('a', 'g') | 'gtgc' |\n","| split | Splits the string at the specified separator, and returns a list | 'hello world'.split() | ['hello', 'world'] |"]},{"cell_type":"markdown","metadata":{"id":"vGPeTv-FPMpY"},"source":["## Recap\n","In this section, we introduced different types of functions: \n","* **built-in**\n","* **object-specific**\n","\n","We also walked through some examples of very practical tasks, using these functions. Another type of function is user-defined function, which we do not cover in this workshop. "]},{"cell_type":"markdown","source":["---\n","\n","# Section 5: Packages (10 min)"],"metadata":{"id":"cqpM19zK7yqo"}},{"cell_type":"markdown","source":["A Python package contains a collection of pre-defined scripts for specific tasks. It allows us to directly use these scripts to accomplish a task of interest, without having to write everything from scratch.\n","\n","We need to install a Python package if it is not already present. [pip](https://pip.pypa.io/en/stable/) is the package installer for Python. To install a package, we could use `!pip install package_name` command. For example, to install [scanpy](https://scanpy.readthedocs.io/en/stable/), a popular package for single-cell RNAseq analysis, we could use `!pip install scanpy`.\n","\n","All major Python libraries are already installed on Colab, so we do not need to install any packages for now. However, we need to `import` a package before we could use it - this import step is required everytime we initiate a new Python environment. As a result, we usually place these `import` codes at the beginning of a script. \n","\n","Sometimes, we name an alias for a package, using the `import package_name as alias` syntax. This way, we just need to use the alias when citing a function from the package, which is convenient if we use the package often. For some popular packages, people set some conventions on what alias to use."],"metadata":{"id":"igpx5q9R76X5"}},{"cell_type":"code","source":["# error when using numpy package without importing first\n","numpy.array([2, 3, 4, 5]) + numpy.array([1, 10, 100, 1000])"],"metadata":{"id":"IjBXl5Du-03j"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# import numpy library, and name it as np\n","import numpy as np"],"metadata":{"id":"FnvVgzki8x23"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# use numpy package after importing (note: need to use np, instead of numpy, because np is the alias we set earlier)\n","np.array([2, 3, 4, 5]) + np.array([1, 10, 100, 1000])"],"metadata":{"id":"qBCxhBJJ-HBY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["---\n","\n","# Section 6: Data manipulation (30 min)"],"metadata":{"id":"LrOFZCDKFH5m"}},{"cell_type":"markdown","source":["Data manipulation is the process in which we organize and clean data to a required format, mostly for downstream analysis. In Python, [pandas](https://pandas.pydata.org/) is a powerful package to perform data manipulation. We will touch upon some basic usage of this package.\n","\n","First, we need to load our data to the cloud server. Let's upload a tsv file, using the upload button on the left panel. We should see the file in our directory (another folder that Colab auto-generates is `sample_data`).\n","\n","> Note: We need to upload the data every time we start a notebook. This is because the cloud server erases the data when the previous notebook is disconnected. There are other ways to connect to data (e.g. mounting the Google Drive), but we will not cover them in this workshop.\n","\n","Once the data is loaded, we can read the data into the Python environment, using the [`read_csv`](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) function from pandas. By default, the function assumes the file to be comma-separated (`,`). However, our file is tab-separated (`\\t`). So we need to add an additional argument, `sep`, to indicate that. "],"metadata":{"id":"CezJj57nI8qS"}},{"cell_type":"code","source":["# import pandas library, and name it as pd \n","import pandas as pd"],"metadata":{"id":"rkHtOFc1FQ-A"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# create a variable called \"data\", and read in data using the read_csv function\n","data = pd.read_csv('functional_analysis.tsv', sep='\\t')"],"metadata":{"id":"OXxaqIOVRIik"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Our data is a two-dimensional dataframe - each row represents a functional process, and each column defines a feature. Below, we will show how to inspect this dataframe and extract data of interest."],"metadata":{"id":"SWZmTDPCMHN2"}},{"cell_type":"markdown","source":["The `head` function prints out the first few rows of the dataframe. This is a helpful way to glimpse the structure of the data, without having to print out all the data."],"metadata":{"id":"gewD8kLtPgTd"}},{"cell_type":"code","source":["# print out the first few rows of the data\n","data.head()"],"metadata":{"id":"fPRajb2AJvPE"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The `shape` function prints out the number of rows and the number of columns for the dataframe. The `columns` function prints out all column names."],"metadata":{"id":"Mf0vvneyTD5A"}},{"cell_type":"code","source":["# print out the shape of the data (number of rows, number of columns)\n","data.shape"],"metadata":{"id":"8_9wrtjaMM5C"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# print out the column names of the data\n","data.columns"],"metadata":{"id":"oMjnvOS8M4Xq"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now that we are familiar with the data, how do we extract specific rows or columns of interest? \n","\n","The first method is to select rows or columns based on **position**. We use the [iloc](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.iloc.html) function from pandas. iloc accepts an integer, a list of integers, or a slice objects. "],"metadata":{"id":"GCk3aGQTNSw9"}},{"cell_type":"code","source":["# select the 4th row\n","data.iloc[3]"],"metadata":{"id":"3JPQCZVmNNCe"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# select the 4th and 5th row. Note that the slicing syntax we introduced in earlier section applies here as well.\n","data.iloc[3:5]"],"metadata":{"id":"DDsRjMThNfhg"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# select the 4th column\n","data.iloc[:,3]"],"metadata":{"id":"NfNZ8pxSNnUO"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# select the 4th and 5th column\n","data.iloc[:,3:5]"],"metadata":{"id":"rtvA_eEWNu-Q"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# select data from 4th and 6th row, 1st and 3rd column\n","data.iloc[[3,5], [0,2]]"],"metadata":{"id":"pH0Z8pCTN7kP"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The second method is to select based on **column names** or **boolean arrays**. We use the [loc](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.loc.html) function from pandas."],"metadata":{"id":"M4l5TcnmQfv6"}},{"cell_type":"code","source":["# select the column p.value\n","data.loc[:,'p.value']"],"metadata":{"id":"XMFPPeIBOIn8"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# select mutiple columns: term.id, term.name, p.value, query.size, overlap.size\n","data.loc[:,['term.id', 'term.name', 'p.value', 'query.size', 'overlap.size']]"],"metadata":{"id":"Zu03rVkhQTMb"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# select where column \"domain\" is \"BP\"\n","idx = data['domain'] == 'BP'\n","data[idx]"],"metadata":{"id":"pJe95C5RQlx-"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Lastly, we use the [sort_values](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html) function to sort values based on a column."],"metadata":{"id":"oWpbKpwf73sH"}},{"cell_type":"code","source":["# sort value based on p.value\n","data.sort_values(by='p.value')"],"metadata":{"id":"p0Lu7pdmTsBL"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["---\n","# Section 7: Visualization (20 min)"],"metadata":{"id":"urZRolgXbsWE"}},{"cell_type":"markdown","source":["Data visualization is an integral part of data analysis. In Python, there are multiple tools for data visualization. For example, pandas has its own [plot](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.plot.html) function. Another popular visualization library is [Matplotlib](https://matplotlib.org/). In this lesson, we will demonstrate [seaborn](https://seaborn.pydata.org/), a library based on Matplotlib, but with extended features for informative graphics. \n","\n","Let's first import matplotlib and seaborn libraries, and load a metadata file that we want to plot some metrics from."],"metadata":{"id":"_MGu0T_TpWuV"}},{"cell_type":"code","source":["# import matplotlib.pyplot and seaborn library\n","import matplotlib.pyplot as plt\n","import seaborn as sns"],"metadata":{"id":"SPFz2l0R_VID"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# read in metadata as a dataframe. Use index_col to indicate which column to be used as the row name\n","metadata = pd.read_csv('metadata.csv', index_col=0)"],"metadata":{"id":"BaXiENxHbwCp"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["metadata"],"metadata":{"id":"-AJbHNnxdQjH"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The [scatterplot](https://seaborn.pydata.org/generated/seaborn.scatterplot.html) function from seaborn plots the relationship between two variables. We add an additional argument, `hue`, to indicate the grouping variable. We add `plt.show()` command, to print out the plot."],"metadata":{"id":"7dV6IHdbq_8t"}},{"cell_type":"code","source":["# plot a scatterplot with \"samplemeans\" as y axis and \"age_in_days\" as x axis. Separate the \"celltype\" with different color.\n","sns.scatterplot(x = 'age_in_days', y = 'samplemeans', hue='celltype', data = metadata)\n","plt.show()"],"metadata":{"id":"9pFrk7NneX1c"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Similarly, we could create [boxplot](https://seaborn.pydata.org/generated/seaborn.boxplot.html) and [violinplot](https://seaborn.pydata.org/generated/seaborn.violinplot.html)."],"metadata":{"id":"YUNqgmD8tEpM"}},{"cell_type":"code","source":["# plot a boxplot with \"samplemeans\" as y axis and \"genotype\" as x axis. Separate the \"celltype\" with different color.\n","sns.boxplot(x = 'genotype', y = 'samplemeans', hue='celltype', data = metadata)\n","plt.show()"],"metadata":{"id":"CTMbYchIsLiU"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# plot a violinplot with \"samplemeans\" as y axis and \"replicate\" as x axis. Separate the \"celltype\" with different color.\n","sns.violinplot(x = 'replicate', y = 'samplemeans', hue=\"celltype\", data = metadata)\n","plt.show()"],"metadata":{"id":"wbhqN0SS4VqX"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["# Section 8: Final exercise\n"],"metadata":{"id":"v7sSgqRZ7ToV"}},{"cell_type":"markdown","source":["In the final exercise, we want to analyze the tumor purity (the percentage of non-cancerous cells present in a tumor sample) across different cancer types. The data is downloaded from this [paper](https://www.nature.com/articles/ncomms9971). Let's first read in the data, and check what the data is about."],"metadata":{"id":"Ktcch00fAeyv"}},{"cell_type":"code","source":["# read in the tumor purity data, and name the variable \"purity\"\n","purity = pd.read_csv(\"tumor_purity.csv\")"],"metadata":{"id":"IUBQoi_N-Nql"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Check the first few entries of the data\n","purity.head()"],"metadata":{"id":"-px6AZxx-ooP"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The data has four columns: Sample_ID, Cancer_type, IHC, and CPE. IHC refers to immunohistochemistry, an experimental method to measure tumor purity. CPE refers to consensus purity estimations, a computational method described in the paper to measure tumor purity."],"metadata":{"id":"wfBBEMsGEYsO"}},{"cell_type":"markdown","source":["1. How many samples are included in this data? Among them, how many are GBM cancer type?"],"metadata":{"id":"6aZjkiO9FJ-H"}},{"cell_type":"code","source":["# Check the size of the data\n","purity.shape"],"metadata":{"id":"c3faS6zF-wKR"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Check the number of samples for GBM\n","sum(purity['Cancer_type']=='GBM')"],"metadata":{"id":"6yKjA8GwIUlM"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["2. The IHC and CPE values are current between 0 and 1. For these two columns, change the value to percentage, by multiplying the current value with 100. Overwrite the old value wiht the new value in these columns."],"metadata":{"id":"kSTe0oxHLV7h"}},{"cell_type":"code","source":["# Change the value to percentage\n","purity.loc[:,'IHC'] = purity.loc[:,'IHC'] * 100\n","purity.loc[:,'CPE'] = purity.loc[:,'CPE'] * 100"],"metadata":{"id":"8_RUKpRTORjI"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["3. Plot a violinplot, with \"Cancer_type\" as the x axis and \"CPE\" as the y axis. "],"metadata":{"id":"pv5b7b0SMYfI"}},{"cell_type":"code","source":["# violinplot\n","sns.violinplot(x = 'Cancer_type', y = 'CPE', data = purity)\n","plt.show()"],"metadata":{"id":"hh-LxDy-ASHi"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["4. Let's make a few more modifications, to make the plot clearer:\n","\n","- Rotate the cancer type labels to 45 degree, so that they don't overlap. Use `xticks` function from Matplotlib, with this [reference solution](https://stackoverflow.com/questions/10998621/rotate-axis-text-in-python-matplotlib).\n","- Remove the x label, and revise the y label to \"Tumor purity (%)\", using `xlabel` and `ylabel` function from Matplotlib. [Here](https://matplotlib.org/3.1.1/gallery/pyplots/pyplot_simple.html#sphx-glr-gallery-pyplots-pyplot-simple-py) is a simple example."],"metadata":{"id":"AAWqSgJ5DiED"}},{"cell_type":"code","source":["sns.violinplot(x = 'Cancer_type', y = 'CPE', data = purity)\n","plt.xticks(rotation=45)\n","plt.xlabel('')\n","plt.ylabel('Tumor purity (%)')\n","plt.show()"],"metadata":{"id":"-1s3IW7zDJfT"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["5 (Optional): The plot is clear now, but it would be even better if we sort the cancer types based on the mean value of tumor purity.\n","To achieve that, we need to:\n","- calculate the mean of purity grouped by each cancer type. Two useful functions for this step are [groupby](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html) and [mean](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html).\n","- sort the purity from lowest to highest, and extract cancer type names\n","- plot violinplot with the new order"],"metadata":{"id":"We-oYBRfNnDD"}},{"cell_type":"code","source":["# Calculate the mean of the purity, grouped by each cancer type\n","purity_mean = purity.groupby(by=['Cancer_type'])['CPE'].mean()"],"metadata":{"id":"cb3IIudRpBKE"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Sort the purity, from lowest to highest\n","purity_sort = purity_mean.sort_values()\n","# Extract the index (cancer type name) as our plotting order\n","new_order = purity_sort.index"],"metadata":{"id":"zUBcS57frdJK"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Re-generate the violin plot, by specifying the new order\n","sns.violinplot(x = 'Cancer_type', y = 'CPE', data = purity, order = new_order)\n","plt.xticks(rotation=45)\n","plt.xlabel('')\n","plt.ylabel('Tumor purity (%)')\n","plt.show()"],"metadata":{"id":"2VTJ4FROsUyt"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"4G7NxBP6nO8s"},"source":["---\n","\n","# Section 9: Class survey (5 min)\n","Before we conclude with final remarks, please take some time to complete this class [survey](http://tinyurl.com/hbc-modules). We appreciate your comments and feedbacks.\n","\n","If you have any suggestions or questions in the future, please contact us at [hbctraining (at) hsph.harvard.edu](mailto:hbctraining@hsph.harvard.edu)."]},{"cell_type":"markdown","metadata":{"id":"UdUoT-1-PMp0"},"source":["---\n","\n","# Section 10: Final remarks (5 min)"]},{"cell_type":"markdown","metadata":{"id":"JDJ5twWF7cRO"},"source":["## Installing Python\n","In this workshop, we used [Google Colab](https://colab.research.google.com/notebooks/intro.ipynb), an online cloud-based notebook. Colab allows us to easily write Python code and execute it just using our internet browser. It is rapidly gaining popularity for collaborative research and teaching (so, you may encounter it again soon!).\n","\n","If you would like to use Python offline though, you will need to install Python on your computer. Python can be downloaded and installed from [here](https://www.python.org/downloads/). The latest release version is Python 3.10 (as of August 2022), and it is recommended that you install the latest version.\n","\n","After Python is installed, there are multiple platforms where you can run Python code. The choice of the platform is a personal preference, and it does not affect code running - at the core we use the native Python environment. Below we introduce some other popular platforms (from among many other options):\n","1. Jupyter Notebook/JupyterLab: [Jupyter Notebook](https://jupyter.org/) is an open-source web application for interactive computing. It allows users to create and share documents that contain code, equations, plots and narrative text. Since Colab is based on the Jupyter notebook environment, we have checked it out in this workshop already. The extension of the notebook file is always `.ipynb`. JupyterLab is a next-generation version of Jupyter Notebook, just with some improvements. You can install them from [here](https://jupyter.org/install).\n","2. Spyder: [Spyder](https://www.spyder-ide.org/) is an open-source integrated development environment (IDE) for scientific programming in Python. It offers a combination of script editing, data analysis, debugging, and visualization. If you are already used to coding in an IDE interface, like RStudio or Matlab, you will find that using Spyder is very familiar and intuitive.\n","3. **(Recommended)** Anaconda: many Python users would use a one-stop-shop package manager like [Anaconda](https://docs.anaconda.com/anaconda/user-guide/getting-started/). Just to note, Anaconda allows users to launch applications and manage conda packages, environments and channels without using command-line commands. After installing Anaconda, you can access its desktop graphical user interface (GUI), the Anaconda Navigator. Applications like Jupyter Notebook, JupyterLab, and Spyder are by default installed and are available on it. Anaconda for Python can be downloaded from [here](https://www.anaconda.com/distribution/)."]},{"cell_type":"markdown","metadata":{"id":"ytQEWGVd7kEw"},"source":["## Python vs R\n","Both Python and R are very popular programming languages, with ample training materials and community support. There are many online discussions about which one would be better for [bioinformatics](https://www.reddit.com/r/bioinformatics/comments/af7wjv/r_language_vs_python_which_is_the_most_necessary/), or for [data analysis](https://www.datacamp.com/community/tutorials/r-or-python-for-data-analysis), in general. We should keep in mind that both have strengths and weaknesses - one language is not better than the other. It all depends on what your use case is and which tools/packages within each language are available to best handle your task. Python is a general purpose programming language with easy-to-understand syntax and is a good start for learning basic programming. R is widely used among stasticians and has field-specific advantages, such as for [RNA-seq analysis](https://bioconductor.org/packages/release/bioc/html/DESeq2.html) or for [data visualization](https://ggplot2.tidyverse.org/index.html). "]},{"cell_type":"markdown","metadata":{"id":"i5xIeWN67nL6"},"source":["## Future learning\n","If you are interested in learning more about basics of Python programming, we listed a few additional resources below:\n","- [Python course on kaggle](https://www.kaggle.com/learn/python)\n","- [Python course on codecademy](https://www.codecademy.com/learn/learn-python)\n","- [Python course on software carpentry](https://swcarpentry.github.io/python-novice-inflammation/)\n","- [A Byte of Python](https://python.swaroopch.com/)\n","- [Python for Biologists](http://userpages.fu-berlin.de/digga/p4b.pdf)"]},{"cell_type":"markdown","metadata":{"id":"BElL7MnQE69F"},"source":["\n","---\n","\n","**Authors**: Jihe Liu, Radhika Khetani\n","\n","*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*"]}]}